docs/design/osd-topology-spread-constraints.md

```diff
@@ -0,0 +1,104 @@
+specific failure domain. This presents two problems:
+
+1. AntiAffinity breaks down when there are more Pods than available Nodes. As
+   soon as this happens, the Kubernetes scheduler acts as if the AntiAffinity
+   doesn't exist, which means that an even distribution of Pods can not be
+   guaranteed.
+2. OSD Pods are locked to only three failure domains, even when more than three
+   exist.
+
+### Goals
+
+Provide an even-as-possible distribution of OSD Pods across all available
+failure domains based on Pod count as opposed to the default of resource
+availability.
+
+### Non-Goals
+
+Always maintain an even distribution of OSD Pods. OSDs can scale at a rate that
+does not match the number of Nodes in a given failure domain, as such there will
+sometimes be uneven distribution of OSD Pods across Nodes in a given failure domain.
+
+## Proposal
+
+Implement the use of Topology Spread Constraints for Ceph OSD Pods. Include the
+ability to specify them in the `StorageCluster.StorageDeviceSet` spec so they
+can be propogated to the resulting `CephCluster`'s SCDSes, providing sane
+defaults if they are not specified.
+
+### Risks and Mitigations
+
+There is an issue in upgrade on the Rook-Ceph side that needs to be addressed.
+The issue is that we want to change the generation of SCDSes from multiple sets
+per StorageDeviceSet to one. This would cause a change in the name generation
+for OSD Pods, which would be a problem in upgrade scenarios going from three
+SCDSes to one.
+
+The initial reason for having multiple StorageClassDeviceSets was to force an
+even distribution of OSD Pods across failure domains (initially, AWS
+Availability Zones (AZs)), since we could not achieve this with Node and Pod
+affinities. Topology Spread Constraints removes the need for this hack.
+
+We will mitigate this by making an initial patch to partially implement
+Topology Spread Constraints in ocs-operator while still maintaining backwards
+compatibility with Rook-Ceph until the upgrade issue can be addressed.
+
+
+## Design Details
+
+The changes will come in two phases:
+
+**Phase 1**
+
+Change the default Placement of StorageDeviceSets to only use Topology Spread
+Constraints, but maintain the split to three StorageClassDeviceSets in the
+generated CephCluster. This preserves the naming scheme for the OSD Pods that
+Rook-Ceph currently implements.
+
+Since we will no longer need to explicitly distinguish between exact failure
+domains, the constraints themselves would be identical across all SCDSes. This
+means that all Pods will be schedulable across all failure domains, regardless
+of which SCDS it belongs to.
+
+For upgrade scenarios, existing OSD Pods would still remain with the old
+Placement configurations. However, because all OSD Pods have appropriate app
+labels to identify them, they would still influence any new OSD Pods created
+with Topology Spread Constraints, allowing the new OSD Pods to be scheduled
+in a more even pattern across failure domains.
```

Comment:
How about adding some specifics about the placement options that will be applied consistently across the three device sets:
- Topology spread constraints for OSD Prepare jobs across the desired failure domain (e.g. `zone` if 3+ zones are available, otherwise `host`)
- Topology spread constraints for OSD daemons across hosts (OSDs will already be tied to a failure domain based on their PV affinity based on where the OSD prepare job ran)
- Node affinity to the nodes selected by the OCS admin

Anything else?

Reply:
Done!